src/thread.cr

```diff
@@ -135,4 +135,58 @@ class Thread
       Fiber.inactive(fiber)
     end
   end
+
+  # :nodoc:
+  def stack : Slice(Void)
+    bottom = Pointer(Void).null
+    top = Pointer(Void).null
+    {% if flag?(:darwin) %}
+      bottom = LibC.pthread_get_stackaddr_np(@th);
+      stack_size = LibC.pthread_get_stacksize_np(@th);
+      top = Pointer(Void).new(bottom.address - stack_size)
+    {% elsif flag?(:freebsd) %}
+      err = LibC.pthread_attr_init(out attr)
```

Comment:
- Pedantic: can we harmonize the implementation within thread? I use `ret` (instead of `err`) and inline the check+raise.
- Leak: if any further call to pthread fails, the `pthread_attr_t` object is never destroyed (this applies to linux as well).

For example:
```crystal
ret = LibC.pthread_attr_init(out attr)
raise Errno.new("pthread_attr_init", ret) unless ret == 0

begin
  ret = LibC.pthread_attr_get_np(@th, pointerof(attr))
  raise Errno.new("pthread_attr_get_np", ret) unless ret == 0

  # ...
ensure
  LibC.pthread_attr_destroy(pointerof(attr))
end
```

Reply:
True enough about that leak.  I had seen the same leak also in #6944 in the changes to `ConditionVariable#initialize` and had assumed that you hadn't used full care to release the resources because if these calls are failing the house is on fire, and maybe the leaky faucet isn't so worrisome. I'll fix this here and submit a separate PR for that other one.

Reply:
Oh, I totally missed it! I only overlooked some possible errnos that are very unlikely (per the manpages) and not worth raising.

Reply:
For the requested changes:

1. On Darwin, on the main thread, we fall back to `getrlimit`.
2. Switched to `ret` not `err` and tried to match nearby code style (`throw ... unless ret == 0`)
3. This was the biggest change.  For platforms that do not have an equivalent of `pthread_getattr_np` we need to cache the thread attributes at the time of creation to access the guard size.  The stack is returned as a separate `Thread::Stack` structure that maintains a pointer, a size, and an optional guard size, which is `nil` if is unknown (i.e. on the main process thread).  If you think this is pedantic, we can simply report a 0 instead.  My previous use of `Slice(Void)` to represent a stack turned out to be a problem because it uses an `Int32` to represent its count, and (rarely, but in principle) a stack can have more bytes than that.

The leak was addressed as well.  For consistency in the code base, probably `ConditionVariable` also ought to get fixed; it doesn't look like the errno's that could occur there are much less likely than the errno's addressed here.